src/projection.jl

```diff
@@ -52,93 +128,197 @@ function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
 end
 
 # Generic
-(project::ProjectTo)(dx::AbstractThunk) = project(unthunk(dx))
 (::ProjectTo{T})(dx::T) where {T} = dx  # not always true, but we can special case for when it isn't
-(::ProjectTo{T})(dx::AbstractZero) where {T} = zero(T)
+(::ProjectTo{T})(dx::AbstractZero) where {T} = dx # zero(T)
+
+# Thunks
+(project::ProjectTo)(dx::Thunk) = Thunk(project ∘ dx.f)
+(project::ProjectTo)(dx::InplaceableThunk) = Thunk(project ∘ dx.val.f) # can't update in-place part
+(project::ProjectTo)(dx::AbstractThunk) = project(unthunk(dx))
 
-# Number
-ProjectTo(::T) where {T<:Number} = ProjectTo{T}()
+# Non-differentiable -- not so sure this is necessary. Keeping value like this is a bit awkward.
```

Comment:
I'm not happy this is done right at all. Maybe the non-diff `value` should be stored directly, and no Project struct is needed for these. I didn't completely unravel what you did here. 

I did run into this but wasn't sure what was failing:
```
julia> x2 = LinearAlgebra.Bidiagonal(rand(4,4), :U)
4×4 Bidiagonal{Float64, Vector{Float64}}:
 0.262634  0.612692    ⋅         ⋅ 
  ⋅        0.0695239  0.564799   ⋅ 
  ⋅         ⋅         0.651733  0.532899
  ⋅         ⋅          ⋅        0.726398

julia> backing(x2)
(dv = [0.26263447890600267, 0.06952385553435525, 0.6517325692754139, 0.7263978804448836], ev = [0.6126916011613305, 0.5647985751765561, 0.5328993346736189], uplo = 'U')

julia> x3 = LinearAlgebra.Tridiagonal(rand(4,4))
4×4 Tridiagonal{Float64, Vector{Float64}}:
 0.591398  0.626837   ⋅         ⋅ 
 0.630986  0.81634   0.76011    ⋅ 
  ⋅        0.310646  0.989977  0.141502
  ⋅         ⋅        0.72248   0.928005

julia> backing(x3)
ERROR: UndefRefError: access to undefined reference
Stacktrace:
 [1] macro expansion
   @ ~/.julia/dev/ChainRulesCore/src/differentials/composite.jl:0 [inlined]
 [2] backing(x::Tridiagonal{Float64, Vector{Float64}})
   @ ChainRulesCore ~/.julia/dev/ChainRulesCore/src/differentials/composite.jl:137
```

Reply:
oh, it is for keeping other data when recursively projecting all fields when doing structs?
This makes a case for not doing that doesn't it.
Maybe we should only go around projecting fields that are `<:AbstractArray` ?

Reply:
The generic recursive `ProjectTo(x)` method blindly calls ProjectTo on all fields. I think that'll error right now if a field were a tuple, but yes, the idea of this thing is still to save things like `uplo::Symbol` somewhere.

It does need to default to ProjectTo, to recurse into `Wrap{Wrap2{Array` right? But perhaps it should not do this for known non-diff types. Is there an existing function which knows this? So that instead of `map(ProjectTo, fields_nt)` it can do
```
map(fields_nt) do field
   is_not_diff(field) ? field : ProjectTo(field)
end
```
And then it's easy for the reconstructor to call something like 
```
maybe_call(f::ProjectTo, x) = f(x)
maybe_call(f, x) = f
construct(T, map(maybe_call, sub_projects, sub_dxs))
```